### ISSUE TITLE
pyopenssl.WrappedSocket and secureransport.WrappedSocket close implementation broken

### ISSUE THREAD

**Issue by graingert** [urllib3/urllib3#2970]
### Subject

calling `ssl_sock.close()` with a makefile open doesn't close the socket when the makefile closes, resulting in a ResourceWarning:

```python
ssl_sock = context.wrap_socket(sock)
f = ssl_sock.makefile()
ssl_sock.close()
f.close() # ssl_sock is still open
del f
del ssl_sock  # ResourceWarning raised here
```

additionally calling `ssl_sock.close()` twice with a makefile open causes the socket to close early

```python
with context.wrap_socket(sock) as ssl_sock:
    with ssl_sock.makefile() as f:
        ssl_sock.close()
        ssl_sock.close()  # socket closed here incorrectly
```

**Comment by graingert** [comment]
the issue is caused by the close methods incorrectly dec-refing _io_refs

https://github.com/urllib3/urllib3/blob/f8fada5ed479e0e6b8440c819632b3a8c040cf65/src/urllib3/contrib/pyopenssl.py#L382-L390
https://github.com/urllib3/urllib3/blob/f8fada5ed479e0e6b8440c819632b3a8c040cf65/src/urllib3/contrib/securetransport.py#L630-L647

the implementation should be as in cpython:

https://github.com/python/cpython/blob/4e04393b2f64f9e44aefa436cbdac42db98d4ef0/Lib/socket.py#L496-L504
```python
def _real_close(self, _ss=_socket.socket):
        # This function should not reference any globals. See issue https://github.com/python/cpython/issues/808164.
        _ss.close(self)

    def close(self):
        # This function should not reference any globals. See issue https://github.com/python/cpython/issues/808164.
        self._closed = True
        if self._io_refs <= 0:
            self._real_close()
```

**Comment by graingert** [comment]
I have a fix in #2842 to fix the ResourceWarning

https://github.com/urllib3/urllib3/blob/45cd5c88677f93cfdb2da5ccd9d631c004661fb1/src/urllib3/contrib/securetransport.py#L631-L650
https://github.com/urllib3/urllib3/blob/45cd5c88677f93cfdb2da5ccd9d631c004661fb1/src/urllib3/contrib/pyopenssl.py#L382-L391

but it should probably be backported to v1
